libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -346,6 +346,35 @@ string YulUtilFunctions::overflowCheckedUIntAddFunction(size_t _bits)
 	});
 }
 
+string YulUtilFunctions::overflowCheckedUIntMulFunction(size_t _bits)
+{
+	solAssert(0 < _bits && _bits <= 256 && _bits % 8 == 0, "");
+	string functionName = "checked_mul_uint_" + to_string(_bits);
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return
+			Whiskers(R"(
+			function <functionName>(x, y) -> product {
+				switch x
+				case 0 { product := 0 }
+				default {
+					if lt(div(<mask>, x), y) { revert(0, 0) }
```

Comment:
Is that check even correct?
```
uint8 x = 0xff;
uint8 y = 1;
return x * y;
```

Reply:
Ah sorry,`lt`!

Reply:
Ok, the proof for `mask / x >= y    =>    x * y is in range` is clear.
Thinking about the converse now...

Reply:
If ``x*y`` is in range, then ``x*y <= mask``, hence ``y <= mask/x``, hence ``mask/x >= y``, so ``!(mask/x < y)``... one has to be careful about division by zero, but that's excluded in the outer switch...

Reply:
(since ``div`` by zero is zero, ``x=0`` would always revert, instead of never reverting, but the other check doesn't work for ``x=0`` either, so it has to be excluded unless we find something entirely different)